tree/ntuple/v7/src/RPagePool.cxx

```diff
@@ -68,30 +96,20 @@ void ROOT::Experimental::Internal::RPagePool::ReleasePage(const RPage &page)
    auto itrLookup = fLookupByBuffer.find(page.GetBuffer());
    assert(itrLookup != fLookupByBuffer.end());
    const auto idx = itrLookup->second;
-   const auto N = fEntries.size();
-   assert(idx < N);
 
    assert(fEntries[idx].fRefCounter >= 1);
    if (--fEntries[idx].fRefCounter == 0) {
-      fLookupByBuffer.erase(itrLookup);
-
-      auto itrPageSet = fLookupByKey.find(fEntries[idx].fKey);
-      assert(itrPageSet != fLookupByKey.end());
-      itrPageSet->second.erase(RPagePosition(page));
-      if (itrPageSet->second.empty())
-         fLookupByKey.erase(itrPageSet);
-
-      if (idx != (N - 1)) {
-         fLookupByBuffer[fEntries[N - 1].fPage.GetBuffer()] = idx;
-         itrPageSet = fLookupByKey.find(fEntries[N - 1].fKey);
-         assert(itrPageSet != fLookupByKey.end());
-         auto itrEntryIdx = itrPageSet->second.find(RPagePosition(fEntries[N - 1].fPage));
-         assert(itrEntryIdx != itrPageSet->second.end());
-         itrEntryIdx->second = idx;
-         fEntries[idx] = std::move(fEntries[N - 1]);
-      }
+      ErasePage(idx, itrLookup);
+   }
+}
 
-      fEntries.resize(N - 1);
+void ROOT::Experimental::Internal::RPagePool::RemoveFromUnusedPages(const RPage &page)
+{
+   auto itr = fUnusedPages.find(page.GetClusterInfo().GetId());
+   if (itr != fUnusedPages.end()) {
+      itr->second.erase(page.GetBuffer());
+      if (itr->second.empty())
+         fUnusedPages.erase(itr);
```

Comment:
Shouldn't we always find the unused page since we add all preloaded pages?

Reply:
We only do this in IMT mode.